src/lua/xml.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "lua/xml.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <lua.h>
+#include <lauxlib.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "lua/utils.h"
+
+int
+luaT_xml_decode(struct lua_State *L)
```

Comment:
Can we onboard luarapidxml (https://github.com/tarantool/luarapidxml) module?

We have json/yaml serializers for now. If we will have built-in xml parser it could be useful in some cases.

Reply:
I was aware of the luarapidxml module, but I decided not to use it, because:
1. Onboarding the rapidxml library to parse XML documents returned by `malloc_info()` looks like an overkill. The size of `src/lua/xml.c` added by this PR is less than the Lua wrapper around the rapidxml library alone.
2. The output format of the `luarapidxml` is too verbose for my goal. I don't want `box.malloc.internal.info()` to have `tag` and `attr` keys for each entry. Compare:
   <details><summary>internal.xml</summary>
   
   ```
   tarantool> box.malloc.internal.info()
   ---
   - malloc:
     - heap:
       - nr: '0'
         system:
         - type: current
           size: '2523136'
         - type: max
           size: '2523136'
         sizes:
         - size:
           - count: '2'
             from: '33'
             to: '33'
             total: '66'
           - count: '5'
             from: '49'
             to: '49'
             total: '245'
           - count: '1'
             from: '193'
             to: '193'
             total: '193'
           - count: '1'
             from: '289'
             to: '289'
             total: '289'
           - count: '1'
             from: '18081'
             to: '18081'
             total: '18081'
           - count: '1'
             from: '50865'
             to: '50865'
             total: '50865'
           unsorted:
           - count: '1'
             from: '4337'
             to: '4337'
             total: '4337'
         total:
         - type: fast
           count: '0'
           size: '0'
         - type: rest
           count: '13'
           size: '594236'
         aspace:
         - type: total
           size: '2523136'
         - type: mprotect
           size: '2523136'
       version: '1'
       system:
       - type: current
         size: '2523136'
       - type: max
         size: '2523136'
       total:
       - type: fast
         count: '0'
         size: '0'
       - type: rest
         count: '13'
         size: '594236'
       - type: mmap
         count: '0'
         size: '0'
       aspace:
       - type: total
         size: '2523136'
       - type: mprotect
         size: '2523136'
   ...
   ```
   
   </details>
   
   (inspired by https://jsonformatter.org/xml-to-yaml)
   
   <details><summary>rapidxml</summary>
   
   ```
   tarantool> box.malloc.internal.info()
   ---
   - 1:
       1:
         1:
           attr:
             count: '5'
             from: '49'
             to: '49'
             total: '245'
           tag: size
         2:
           attr:
             count: '1'
             from: '193'
             to: '193'
             total: '193'
           tag: size
         3:
           attr:
             count: '1'
             from: '289'
             to: '289'
             total: '289'
           tag: size
         4:
           attr:
             count: '1'
             from: '18081'
             to: '18081'
             total: '18081'
           tag: size
         5:
           attr:
             count: '1'
             from: '50865'
             to: '50865'
             total: '50865'
           tag: size
         6:
           attr:
             count: '1'
             from: '3345'
             to: '3345'
             total: '3345'
           tag: unsorted
         tag: sizes
       2:
         attr:
           type: fast
           count: '0'
           size: '0'
         tag: total
       3:
         attr:
           type: rest
           count: '11'
           size: '593178'
         tag: total
       4:
         attr:
           type: current
           size: '2523136'
         tag: system
       5:
         attr:
           type: max
           size: '2523136'
         tag: system
       6:
         attr:
           type: total
           size: '2523136'
         tag: aspace
       7:
         attr:
           type: mprotect
           size: '2523136'
         tag: aspace
       attr:
         nr: '0'
       tag: heap
     2:
       attr:
         type: fast
         count: '0'
         size: '0'
       tag: total
     3:
       attr:
         type: rest
         count: '11'
         size: '593178'
       tag: total
     4:
       attr:
         type: mmap
         count: '0'
         size: '0'
       tag: total
     5:
       attr:
         type: current
         size: '2523136'
       tag: system
     6:
       attr:
         type: max
         size: '2523136'
       tag: system
     7:
       attr:
         type: total
         size: '2523136'
       tag: aspace
     8:
       attr:
         type: mprotect
         size: '2523136'
       tag: aspace
     attr:
       version: '1'
     tag: malloc
   ...
   ```
   
   </details>

Initially, I planned to parse `malloc_info()` output right in `box/src/lua/slab.cc`, but decided to factor this code out into a separate internal Lua module so that I can test it with `luatest`.

If you need to have an XML parser onboard and have a reason for that, please file a ticket. But even if we decide to add a public XML parser, I'm not sure we'll actually choose `luarapidxml`. There undoubtedly will be an investigation about which XML library suits best for our purposes. Since we needed `malloc_info()` yesterday, I think it's better to add ad hoc internal XML parser for now, just for `malloc_info()` sake. Later on we can either improve it, or replace it with an external library.

Reply:
Ok. Thanks for explanation.